numpyro/contrib/nested_sampling.py

```diff
@@ -0,0 +1,254 @@
+        self._results = None
+
+    def run(self, rng_key, *args, **kwargs):
+        """
+        Run the nested samplers and collect weighted samples.
+
+        :param random.PRNGKey rng_key: Random number generator key to be used for the sampling.
+        :param args: The arguments needed by the `model`.
+        :param kwargs: The keyword arguments needed by the `model`.
+        """
+        rng_sampling, rng_predictive = random.split(rng_key)
+        # reparam the model so that latent sites have Uniform(0, 1) priors
+        prototype_trace = trace(seed(self.model, rng_key)).get_trace(*args, **kwargs)
+        param_names = [site["name"] for site in prototype_trace.values()
+                       if site["type"] == "sample" and not site["is_observed"]
+                       and site["infer"].get("enumerate", "") != "parallel"]
+        deterministics = [site["name"] for site in prototype_trace.values()
+                          if site["type"] == "deterministic"]
+        reparam_model = reparam(self.model, config={k: UniformReparam() for k in param_names})
+
+        # enable enumerate if needed
+        has_enum = any(site["type"] == "sample" and site["infer"].get("enumerate", "") == "parallel"
+                       for site in prototype_trace.values())
+        if has_enum:
+            from numpyro.contrib.funsor import enum, log_density as log_density_
+
+            max_plate_nesting = _guess_max_plate_nesting(prototype_trace)
+            _validate_model(prototype_trace)
+            reparam_model = enum(reparam_model, -max_plate_nesting - 1)
+        else:
+            log_density_ = log_density
+
+        def loglik_fn(**params):
+            return log_density_(reparam_model, args, kwargs, params)[0]
+
+        # use NestedSampler with identity prior chain
+        prior_chain = PriorChain()
+        for name in param_names:
+            prior = UniformPrior(name + "_base", prototype_trace[name]["fn"].shape())
+            prior_chain.push(prior)
+        ns = OrigNestedSampler(loglik_fn, prior_chain, sampler_name=self.sampler_name,
+                               sampler_kwargs={"depth": self.depth, "num_slices": self.num_slices},
+                               max_samples=self.max_samples,
+                               num_live_points=self.num_live_points,
+                               collect_samples=True)
+        results = ns(rng_sampling, termination_frac=self.termination_frac)
+
+        num_samples = int(device_get(results.num_samples))
+        log_weights = results.log_p[:num_samples]
+        base_samples = {k: v[:num_samples] for k, v in results.samples.items()}
+        # filter out samples with -inf weights and invalid values
+        valid_masks = [jnp.isfinite(log_weights)] + \
+            [((v > 0) & (v < 1)).reshape((num_samples, -1)).all(-1)
+             for v in base_samples.values()]
+        mask = jnp.stack(valid_masks, -1).all(-1)
+        self._log_weights = log_weights[mask]
```

Comment:
I saw some `-inf` in log_weights[:num_samples] and in samples[:num_samples] of the example above (in the docstring of NestedSampler), which prevented me from collecting some statistics after sampling. I saw your plotting framework also does some sort of [filtering](https://github.com/Joshuaalbert/jaxns/blob/master/jaxns/plotting.py#L84) after getting `samples[:num_samples]` so I thought that those -inf are expected.

Reply:
Those lines are related to giving samples with `inf` no weight. I.e. it happens that the you feed in a 1 to the normal distribution quantile function you'll get `inf`. In principle, this shouldn't happen because the samplers try to avoid that happening, but the plotting filters those out for safety. In general, we don't need to filter out logL=-inf for plotting.

`log_weights[:num_samples]` shouldn't have any inf. Make sure you use jaxns=0.0.5 as there would have been some in smaller versions.

Reply:
Confirms that those `inf` are gone in jaxns 0.0.5. I'll ping jaxns to `0.0.5` to avoid issues like this.

Reply:
great